csharp/src/Ice/Endpoint.cs

```diff
@@ -247,8 +226,46 @@ protected Endpoint(EndpointData data, Communicator communicator, Protocol protoc
         }
     }
 
-    internal static class EndpointExtensions
+    public static class EndpointExtensions
     {
+        /// <summary>Creates an endpoint from an <see cref="EndpointData"/> struct.</summary>
+        /// <param name="data">The endpoint's data.</param>
+        /// <param name="communicator">The communicator.</param>
+        /// <param name="protocol">The endpoint's protocol. Must be ice2 or greater.</param>
+        /// <returns>A new endpoint.</returns>
+        public static Endpoint ToEndpoint(
+            this EndpointData data,
+            Communicator communicator,
+            Protocol protocol = Protocol.Ice2)
+        {
+            if ((byte)protocol < (byte)Protocol.Ice2)
+            {
+                throw new ArgumentException("protocol must be ice2 or greater", nameof(protocol));
+            }
+
+            Ice2EndpointFactory? factory =
+                    protocol == Protocol.Ice2 ? communicator.FindIce2EndpointFactory(data.Transport) : null;
+
+            return factory?.Invoke(data, communicator) ?? UniversalEndpoint.Create(data, communicator, protocol);
+        }
+
+        /// <summary>Creates an endpoint data list from a sequence of endpoints.</summary>
+        /// <param name="endpoints">The sequence of endpoints.</param>
+        /// <returns>A new list of endpoint data.</returns>
+        public static List<EndpointData> ToEndpointDataList(this IEnumerable<Endpoint> endpoints) =>
```

Comment:
I added these 2 public extensions (ToEndpointDataList and ToEndpointList) is because you typically don't pass a single endpoint but a sequence<Endpoint> in your Slice operation (like the new Locator operations added by this PR), and then these endpoints are actually passed as a sequence<EndpointData> while what you often start with is a list of endpoints (IReadOnlyList<Endpoint>), not a IReadOnlyList<EndpointData>.

So Data / ToEndpoint to translate Endpoint to/from EndpointData are actually not that useful since 99% of the time you're dealing with sequences. 

This method should be used by ObjectAdapter (fixing), and is used by the location test.